Compiling with Javascript tools=yes fails (no exceptions support for Clipper and Assimp build)

**Godot version:**
master (https://github.com/godotengine/godot/commit/2c9536b655d006c697a3cf830ad3352f1a1ed9f2)


**OS/device including version:**
Arch Linux


**Issue description:**
The javascript export (html5) fails with tools=yes.
I think it should work because there was the website http://godot.online.
```
[ 72%] Compiling ==> editor/collada/collada.cpp
thirdparty/misc/clipper.cpp:901:7: error: cannot use 'throw' with exceptions
      disabled
      throw clipperException("Coordinate outside allowed range");
      ^
thirdparty/misc/clipper.cpp:1049:5: error: cannot use 'throw' with exceptions
      disabled
    throw clipperException("AddPath: Open paths must be subject.");
    ^
thirdparty/misc/clipper.cpp:1081:5: error: cannot use 'throw' with exceptions
      disabled
    throw; //range test fails
    ^
thirdparty/misc/clipper.cpp:1065:3: error: cannot use 'try' with exceptions
      disabled
  try
  ^
thirdparty/misc/clipper.cpp:1445:5: error: cannot use 'throw' with exceptions
      disabled
    throw clipperException("UpdateEdgeIntoAEL: invalid call");
    ^
thirdparty/misc/clipper.cpp:1513:5: error: cannot use 'throw' with exceptions
      disabled
    throw clipperException("Error: PolyTree struct is needed for open pa...
    ^
thirdparty/misc/clipper.cpp:1563:3: error: cannot use 'try' with exceptions
      disabled
  try {
  ^
thirdparty/misc/clipper.cpp:2841:5: error: cannot use 'throw' with exceptions
      disabled
    throw clipperException("ProcessIntersections error");
    ^
thirdparty/misc/clipper.cpp:2830:3: error: cannot use 'try' with exceptions
      disabled
  try {
  ^
thirdparty/misc/clipper.cpp:3005:8: error: cannot use 'throw' with exceptions
      disabled
  else throw clipperException("DoMaxima error");
       ^
10 errors generated.
shared:ERROR: compiler frontend failed to generate LLVM bitcode, halting
scons: *** [thirdparty/misc/clipper.javascript.opt.tools.64.bc] Error 1
scons: building terminated because of errors.
```



**Steps to reproduce:**
```scons p=javascript tools=yes target=release_debug bits=64 -j 8```

OS: Windows 10 WSL
Godot version: 33897d9

While trying to resolve this issue and in order to test #29003 I got this error:
```
$ scons platform=javascript target=release_debug tools=yes bits=64 -j4 warnings=extra
scons: Reading SConscript files ...
Using custom build name: Xrayez
WebM SIMD optimizations are disabled. Check if your CPU architecture, CPU bits or platform are supported!
Checking for C header file mntent.h... (cached) yes
scons: done reading SConscript files.
scons: Building targets ...
[  2%] Compiling ==> platform/javascript/os_javascript.cpp
[  5%] Linking Static Library ==> main/libmain.javascript.opt.tools.64.bc
[  5%] Compiling ==> main/tests/test_astar.cpp
[  5%] Compiling ==> main/tests/test_gdscript.cpp
Ranlib Library         ==> main/libmain.javascript.opt.tools.64.bc
[  5%] Compiling ==> main/tests/test_gui.cpp
[  5%] Compiling ==> main/tests/test_main.cpp
platform/javascript/os_javascript.cpp:997:6: error: use of undeclared identifier 'emscripten_sample_gamepad_data'
        if (emscripten_sample_gamepad_data() == EMSCRIPTEN_RESULT_SUCCESS)
            ^
1 error generated.
ERROR:root:compiler frontend failed to generate LLVM bitcode, halting
scons: *** [platform/javascript/os_javascript.javascript.opt.tools.64.bc] Error 1
scons: building terminated because of errors.
```

Ok, I've had an outdated 1.38.4 Emscripten SDK, updated to 1.38.31 now, but I get these errors now:

```
$ scons platform=javascript target=release_debug tools=yes bits=64 debug_symbols=no -j4 warnings=extra
scons: Reading SConscript files ...
Using custom build name: Xrayez
WebM SIMD optimizations are disabled. Check if your CPU architecture, CPU bits or platform are supported!
Checking for C header file mntent.h... yes
scons: done reading SConscript files.
scons: Building targets ...
[  2%] Compiling ==> platform/javascript/os_javascript.cpp
[  5%] Linking Static Library ==> main/libmain.javascript.opt.tools.64.bc
[  5%] Compiling ==> main/tests/test_math.cpp
[  5%] Compiling ==> main/tests/test_oa_hash_map.cpp
Ranlib Library         ==> main/libmain.javascript.opt.tools.64.bc
[  5%] Compiling ==> main/tests/test_ordered_hash_map.cpp
[  5%] Compiling ==> main/tests/test_physics.cpp
[  5%] Compiling ==> main/tests/test_physics_2d.cpp
[  5%] Compiling ==> main/tests/test_render.cpp
[  5%] Compiling ==> main/tests/test_shader_lang.cpp
[  6%] Compiling ==> main/tests/test_string.cpp
[  6%] Compiling ==> modules/register_module_types.gen.cpp
[  7%] Compiling ==> thirdparty/assimp/code/BaseImporter.cpp
[  7%] Compiling ==> thirdparty/assimp/code/BaseProcess.cpp
[  7%] Compiling ==> thirdparty/assimp/code/Bitmap.cpp
[  7%] Linking Static Library ==> main/tests/libtests.javascript.opt.tools.64.bc
In file included from thirdparty/assimp/code/BaseImporter.cpp:48:
In file included from thirdparty/assimp/include/assimp/BaseImporter.h:47:
thirdparty/assimp/include/assimp/Exceptional.h:96:13: error: cannot use 'throw' with exceptions disabled
            throw;
            ^
thirdparty/assimp/include/assimp/Exceptional.h:95:9: error: cannot use 'try' with exceptions disabled
        try {
        ^
In file included from thirdparty/assimp/code/BaseProcess.cpp:46:
In file included from thirdparty/assimp/include/assimp/BaseImporter.h:47:
thirdparty/assimp/include/assimp/Exceptional.h:96:13: error: cannot use 'throw' with exceptions disabled
            throw;
            ^
thirdparty/assimp/include/assimp/Exceptional.h:95:9: error: cannot use 'try' with exceptions disabled
        try {
        ^
In file included from thirdparty/assimp/code/BaseImporter.cpp:50:
In file included from thirdparty/assimp/code/FileSystemFilter.h:51:
thirdparty/assimp/include/assimp/fast_atof.h:190:9: error: cannot use 'throw' with exceptions disabled
        throw std::invalid_argument( std::string( "The string \"" ) + in + "\" cannot be converted into a value." );
        ^
thirdparty/assimp/include/assimp/fast_atof.h:288:9: error: cannot use 'throw' with exceptions disabled
        throw std::invalid_argument("Cannot parse string "
        ^
[  7%] Compiling ==> thirdparty/assimp/code/CInterfaceIOWrapper.cpp
thirdparty/assimp/code/BaseImporter.cpp:99:5: error: cannot use 'try' with exceptions disabled
    try
    ^
In file included from thirdparty/assimp/code/BaseImporter.cpp:323:
In file included from thirdparty/assimp/code/../contrib/utf8cpp/source/utf8.h:31:
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:76:13: error: cannot use 'throw' with exceptions disabled
            throw invalid_code_point(cp);
            ^
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:110:21: error: cannot use 'throw' with exceptions disabled
                    throw not_enough_room();
                    ^
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:145:17: error: cannot use 'throw' with exceptions disabled
                throw not_enough_room();
                ^
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:149:17: error: cannot use 'throw' with exceptions disabled
                throw invalid_utf8(*it);
                ^
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:151:17: error: cannot use 'throw' with exceptions disabled
                throw invalid_code_point(cp);
                ^
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:167:13: error: cannot use 'throw' with exceptions disabled
            throw not_enough_room();
            ^
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:173:17: error: cannot use 'throw' with exceptions disabled
                throw invalid_utf8(*it); // error - no lead byte in the sequence
                ^
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:184:17: error: cannot use 'throw' with exceptions disabled
                throw invalid_utf8(*it); // error - no lead byte in the sequence
                ^
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:218:25: error: cannot use 'throw' with exceptions disabled
                        throw invalid_utf16(static_cast<uint16_t>(trail_surrogate));
                        ^
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:221:21: error: cannot use 'throw' with exceptions disabled
                    throw invalid_utf16(static_cast<uint16_t>(cp));
                    ^
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:226:17: error: cannot use 'throw' with exceptions disabled
                throw invalid_utf16(static_cast<uint16_t>(cp));
                ^
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:280:15: error: cannot use 'throw' with exceptions disabled
              throw std::out_of_range("Invalid utf-8 iterator position");
              ^
thirdparty/assimp/code/../contrib/utf8cpp/source/utf8/checked.h:292:15: error: cannot use 'throw' with exceptions disabled
              throw std::logic_error("Comparing utf-8 iterators defined with different ranges");
              ^
thirdparty/assimp/code/BaseImporter.cpp:331:9: error: cannot use 'throw' with exceptions disabled
        throw DeadlyImportError("File is too small");
        ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
shared:ERROR: compiler frontend failed to generate LLVM bitcode, halting
scons: *** [thirdparty/assimp/code/BaseImporter.javascript.opt.tools.64.bc] Error 1
thirdparty/assimp/code/BaseProcess.cpp:80:5: error: cannot use 'try' with exceptions disabled
    try
    ^
3 errors generated.
shared:ERROR: compiler frontend failed to generate LLVM bitcode, halting
scons: *** [thirdparty/assimp/code/BaseProcess.javascript.opt.tools.64.bc] Error 1
Ranlib Library         ==> main/tests/libtests.javascript.opt.tools.64.bc
scons: building terminated because of errors.
```
@fire seems like your `assimp` module fails too in platforms with disabled exceptions, this needs to be fixed by either disabling assimp for web with tools enabled, or provide similar fixes like I did with Clipper (see above). 😃

--- a/platform/android/detect.py
+++ b/platform/android/detect.py
@@ -214,13 +214,14 @@ def configure(env):
     lib_sysroot = env["ANDROID_NDK_ROOT"] + "/platforms/" + env['ndk_platform'] + "/" + env['ARCH']
 
     ## Compile flags
-
-    if env['android_stl']:
+    # Disable exceptions and rtti on non-tools (template) builds
+    if env['tools'] or env['android_stl']:
         env.Append(CPPFLAGS=["-isystem", env["ANDROID_NDK_ROOT"] + "/sources/cxx-stl/llvm-libc++/include"])
         env.Append(CPPFLAGS=["-isystem", env["ANDROID_NDK_ROOT"] + "/sources/cxx-stl/llvm-libc++abi/include"])
         env.Append(CXXFLAGS=['-frtti', "-std=gnu++14"])
     else:
         env.Append(CXXFLAGS=['-fno-rtti', '-fno-exceptions'])
+        # Don't use dynamic_cast, necessary with no-rtti.
         env.Append(CPPFLAGS=['-DNO_SAFE_CAST'])
 
     ndk_version = get_ndk_version(env["ANDROID_NDK_ROOT"])

--- a/platform/iphone/detect.py
+++ b/platform/iphone/detect.py
@@ -115,10 +115,12 @@ def configure(env):
         env.Append(CPPFLAGS=['-DNEED_LONG_INT'])
         env.Append(CPPFLAGS=['-DLIBYUV_DISABLE_NEON'])
 
-    if env['ios_exceptions']:
-        env.Append(CCFLAGS=['-fexceptions'])
-    else:
-        env.Append(CCFLAGS=['-fno-exceptions'])
+    # Disable exceptions on non-tools (template) builds
+    if not env['tools']:
+        if env['ios_exceptions']:
+            env.Append(CCFLAGS=['-fexceptions'])
+        else:
+            env.Append(CCFLAGS=['-fno-exceptions'])
 
     ## Link flags
 

--- a/platform/javascript/detect.py
+++ b/platform/javascript/detect.py
@@ -110,10 +110,12 @@ def configure(env):
     # once feasible also consider memory buffer size issues.
     env.Append(CPPDEFINES=['NO_THREADS'])
 
-    # These flags help keep the file size down.
-    env.Append(CCFLAGS=['-fno-exceptions', '-fno-rtti'])
-    # Don't use dynamic_cast, necessary with no-rtti.
-    env.Append(CPPDEFINES=['NO_SAFE_CAST'])
+    # Disable exceptions and rtti on non-tools (template) builds
+    if not env['tools']:
+        # These flags help keep the file size down.
+        env.Append(CCFLAGS=['-fno-exceptions', '-fno-rtti'])
+        # Don't use dynamic_cast, necessary with no-rtti.
+        env.Append(CPPDEFINES=['NO_SAFE_CAST'])
 
     if env['javascript_eval']:
         env.Append(CPPDEFINES=['JAVASCRIPT_EVAL_ENABLED'])

